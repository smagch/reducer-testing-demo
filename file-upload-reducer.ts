export type FileInterface = {
  name: string;
  stream: () => ReadableStream<Uint8Array>;
  size: number;
};

type Status =
  | "waiting"
  | "validating"
  | "validationFailed"
  | "validationSucceeded"
  | "uploading"
  | "uploadSucceeded"
  | "uploadFailed";

type FileStats = {
  hash: string;
  rowCount: number;
  columns: string[];
};

export type Item = {
  file: FileInterface;
  progress: number;
  status: Status;
  stats?: FileStats;
  err?: string;
};

export type State = {
  items: Item[];
  validating?: FileInterface;
  uploading?: Item[];
};

export type Action =
  | { type: "APPEND_FILES"; files: FileInterface[] }
  | { type: "VALIDATION_SUCCEEDED"; filename: string; stats: FileStats }
  | { type: "VALIDATION_FAILED"; filename: string; err: Error | string }
  | { type: "PROGRESS"; filename: string; progress: number }
  | { type: "UPLOAD_SUCCEEDED"; filename: string }
  | { type: "REMOVE_FILE"; filename: string };

function updateValidateStatus(state: State): State {
  if (!state.items.length) {
    return state;
  }

  let validatingItem = state.items.find((item) => {
    return item.status === "validating";
  });

  if (!validatingItem) {
    const index = state.items.findIndex((item) => {
      return item.status === "waiting";
    });
    if (index !== -1) {
      const current = state.items[index];
      state.items[index] = {
        ...current,
        status: "validating",
      };
      validatingItem = state.items[index];
    }
  }

  if (validatingItem) {
    state.validating = validatingItem.file;
  }

  const uploadingItems = state.items.filter((item) => {
    return item.status === "uploading";
  });
  const validationSucceededItems = state.items.filter((item) => {
    return item.status === "validationSucceeded";
  });
  if (uploadingItems.length < 3 && validationSucceededItems.length) {
    const pickCount = Math.min(
      3 - uploadingItems.length,
      validationSucceededItems.length
    );
    const itemsToUpload = validationSucceededItems.slice(0, pickCount);
    itemsToUpload.forEach((item) => {
      item.status = "uploading";
    });
    uploadingItems.push(...itemsToUpload);
    state.uploading = uploadingItems;
  }

  return state;
}

export const reducer = (state: State, action: Action): State => {
  switch (action.type) {
    case "APPEND_FILES": {
      const items: Item[] = action.files
        .filter((file) => {
          const exists = state.items.some((item) => {
            return item.file.name === file.name;
          });
          return !exists;
        })
        .map((file) => {
          return {
            file,
            status: "waiting",
            progress: 0,
          };
        });

      return updateValidateStatus({
        uploading: state.uploading,
        items: state.items.concat(items),
      });
    }

    case "PROGRESS": {
      const items: Item[] = state.items.map((item) => {
        if (item.file.name !== action.filename) {
          return item;
        }
        return {
          ...item,
          progress: action.progress,
        };
      });

      return {
        ...state,
        items,
      };
    }

    case "VALIDATION_SUCCEEDED": {
      const items: Item[] = state.items.map((item) => {
        if (item.file.name !== action.filename) {
          return item;
        }
        return {
          ...item,
          status: "validationSucceeded",
          progress: 0,
          stats: action.stats,
        };
      });

      return updateValidateStatus({ items, uploading: state.uploading });
    }

    case "VALIDATION_FAILED": {
      const items: Item[] = state.items.map((item) => {
        if (item.file.name !== action.filename) {
          return item;
        }
        return {
          ...item,
          status: "validationFailed",
          err: action.err instanceof Error ? action.err.message : action.err,
          progress: 0,
        };
      });
      return updateValidateStatus({ items, uploading: state.uploading });
    }

    case "UPLOAD_SUCCEEDED": {
      const items: Item[] = state.items.map((item) => {
        if (item.file.name !== action.filename) {
          return item;
        }
        return {
          ...item,
          status: "uploadSucceeded",
          progress: 1,
        };
      });
      const uploading = (state.uploading?.slice() || []).filter((item) => {
        return item.file.name !== action.filename;
      });

      return updateValidateStatus({ items, uploading });
    }

    case "REMOVE_FILE": {
      const items: Item[] = state.items.filter((item) => {
        return item.file.name !== action.filename;
      });
      return updateValidateStatus({ items, uploading: state.uploading });
    }

    default: {
      return state;
    }
  }
};

export function reset(): State {
  return {
    items: [],
  };
}
