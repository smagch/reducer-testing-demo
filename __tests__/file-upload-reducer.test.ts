import { reducer, reset, State, FileInterface } from "../file-upload-reducer";

const file = (filename: string): FileInterface => {
  const encoder = new TextEncoder();
  const chunk = encoder.encode(filename);
  return {
    name: filename,
    size: chunk.byteLength,
    stream: (): ReadableStream<Uint8Array> => {
      return new ReadableStream<Uint8Array>({
        pull(controller: ReadableStreamDefaultController<Uint8Array>) {
          controller.enqueue(chunk);
          controller.close();
        },
      }) as ReadableStream<Uint8Array>;
    },
  };
};

function getState(): State {
  return {
    items: [
      {
        file: file("file1.txt"),
        progress: 0.1,
        status: "validating" as const,
      },
      {
        file: file("file2.txt"),
        progress: 0,
        status: "waiting" as const,
      },
      {
        file: file("file3.txt"),
        progress: 0,
        status: "waiting" as const,
      },
      {
        file: file("file4.txt"),
        progress: 0,
        status: "waiting" as const,
      },
    ],
  };
}

test("empty state", () => {
  expect(reset()).toMatchSnapshot();
});

test("APPEND_FILES dispatch", () => {
  let state = reducer(reset(), {
    type: "APPEND_FILES",
    files: [file("file1.txt")],
  });

  expect(state).toMatchSnapshot();

  state = reducer(state, {
    type: "APPEND_FILES",
    files: [file("file2.txt"), file("file3.txt")],
  });

  expect(state).toMatchSnapshot();
});

test("VALIDATION_SUCCEEDED", () => {
  let state = getState();
  for (const filename of ["file1.txt", "file2.txt", "file3.txt", "file4.txt"]) {
    state = reducer(state, {
      type: "VALIDATION_SUCCEEDED",
      filename,
      stats: {
        hash: "hash",
        rowCount: 1000,
        columns: ["id", "name", "email"],
      },
    });
    expect(state).toMatchSnapshot();
  }
});

test("VALIDATION_FAILED", () => {
  let state = getState();
  state = reducer(state, {
    type: "VALIDATION_FAILED",
    filename: "file1.txt",
    err: "invalid encoding",
  });
  expect(state).toMatchSnapshot();

  state = reducer(state, {
    type: "VALIDATION_FAILED",
    filename: "file2.txt",
    err: "invalid csv format",
  });
  expect(state).toMatchSnapshot();
});

test("PROGRESS", () => {
  let state: State = {
    items: [
      {
        file: file("file1.txt"),
        progress: 0,
        status: "uploading" as const,
      },
      {
        file: file("file2.txt"),
        progress: 0,
        status: "uploading" as const,
      },
    ],
  };

  state = reducer(state, {
    type: "PROGRESS",
    progress: 0.2,
    filename: "file1.txt",
  });

  expect(state).toMatchSnapshot();
});

test("UPLOAD_SUCCEEDED", () => {
  const items = [
    {
      file: file("file1.txt"),
      progress: 0,
      status: "uploading" as const,
    },
    {
      file: file("file2.txt"),
      progress: 0,
      status: "uploading" as const,
    },
    {
      file: file("file3.txt"),
      progress: 0,
      status: "uploading" as const,
    },
    {
      file: file("file4.txt"),
      progress: 0,
      status: "validationSucceeded" as const,
    },
  ];

  let state: State = {
    items,
    uploading: [items[0], items[1], items[2]],
  };

  state = reducer(state, {
    type: "UPLOAD_SUCCEEDED",
    filename: "file1.txt",
  });

  expect(state).toMatchSnapshot();
});

test("REMOVE FILE", () => {
  let state = getState();
  state = reducer(state, {
    type: "REMOVE_FILE",
    filename: "file2.txt",
  });
  expect(state).toMatchSnapshot();

  state = reducer(state, {
    type: "REMOVE_FILE",
    filename: "file1.txt",
  });
  expect(state).toMatchSnapshot();
});
