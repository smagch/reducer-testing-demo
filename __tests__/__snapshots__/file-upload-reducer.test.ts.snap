// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`APPEND_FILES dispatch 1`] = `
Object {
  "items": Array [
    Object {
      "file": Object {
        "name": "file1.txt",
        "size": 9,
        "stream": [Function],
      },
      "progress": 0,
      "status": "validating",
    },
  ],
  "uploading": undefined,
  "validating": Object {
    "name": "file1.txt",
    "size": 9,
    "stream": [Function],
  },
}
`;

exports[`APPEND_FILES dispatch 2`] = `
Object {
  "items": Array [
    Object {
      "file": Object {
        "name": "file1.txt",
        "size": 9,
        "stream": [Function],
      },
      "progress": 0,
      "status": "validating",
    },
    Object {
      "file": Object {
        "name": "file2.txt",
        "size": 9,
        "stream": [Function],
      },
      "progress": 0,
      "status": "waiting",
    },
    Object {
      "file": Object {
        "name": "file3.txt",
        "size": 9,
        "stream": [Function],
      },
      "progress": 0,
      "status": "waiting",
    },
  ],
  "uploading": undefined,
  "validating": Object {
    "name": "file1.txt",
    "size": 9,
    "stream": [Function],
  },
}
`;

exports[`PROGRESS 1`] = `
Object {
  "items": Array [
    Object {
      "file": Object {
        "name": "file1.txt",
        "size": 9,
        "stream": [Function],
      },
      "progress": 0.2,
      "status": "uploading",
    },
    Object {
      "file": Object {
        "name": "file2.txt",
        "size": 9,
        "stream": [Function],
      },
      "progress": 0,
      "status": "uploading",
    },
  ],
}
`;

exports[`REMOVE FILE 1`] = `
Object {
  "items": Array [
    Object {
      "file": Object {
        "name": "file1.txt",
        "size": 9,
        "stream": [Function],
      },
      "progress": 0.1,
      "status": "validating",
    },
    Object {
      "file": Object {
        "name": "file3.txt",
        "size": 9,
        "stream": [Function],
      },
      "progress": 0,
      "status": "waiting",
    },
    Object {
      "file": Object {
        "name": "file4.txt",
        "size": 9,
        "stream": [Function],
      },
      "progress": 0,
      "status": "waiting",
    },
  ],
  "uploading": undefined,
  "validating": Object {
    "name": "file1.txt",
    "size": 9,
    "stream": [Function],
  },
}
`;

exports[`REMOVE FILE 2`] = `
Object {
  "items": Array [
    Object {
      "file": Object {
        "name": "file3.txt",
        "size": 9,
        "stream": [Function],
      },
      "progress": 0,
      "status": "validating",
    },
    Object {
      "file": Object {
        "name": "file4.txt",
        "size": 9,
        "stream": [Function],
      },
      "progress": 0,
      "status": "waiting",
    },
  ],
  "uploading": undefined,
  "validating": Object {
    "name": "file3.txt",
    "size": 9,
    "stream": [Function],
  },
}
`;

exports[`UPLOAD_SUCCEEDED 1`] = `
Object {
  "items": Array [
    Object {
      "file": Object {
        "name": "file1.txt",
        "size": 9,
        "stream": [Function],
      },
      "progress": 1,
      "status": "uploadSucceeded",
    },
    Object {
      "file": Object {
        "name": "file2.txt",
        "size": 9,
        "stream": [Function],
      },
      "progress": 0,
      "status": "uploading",
    },
    Object {
      "file": Object {
        "name": "file3.txt",
        "size": 9,
        "stream": [Function],
      },
      "progress": 0,
      "status": "uploading",
    },
    Object {
      "file": Object {
        "name": "file4.txt",
        "size": 9,
        "stream": [Function],
      },
      "progress": 0,
      "status": "uploading",
    },
  ],
  "uploading": Array [
    Object {
      "file": Object {
        "name": "file2.txt",
        "size": 9,
        "stream": [Function],
      },
      "progress": 0,
      "status": "uploading",
    },
    Object {
      "file": Object {
        "name": "file3.txt",
        "size": 9,
        "stream": [Function],
      },
      "progress": 0,
      "status": "uploading",
    },
    Object {
      "file": Object {
        "name": "file4.txt",
        "size": 9,
        "stream": [Function],
      },
      "progress": 0,
      "status": "uploading",
    },
  ],
}
`;

exports[`VALIDATION_FAILED 1`] = `
Object {
  "items": Array [
    Object {
      "err": "invalid encoding",
      "file": Object {
        "name": "file1.txt",
        "size": 9,
        "stream": [Function],
      },
      "progress": 0,
      "status": "validationFailed",
    },
    Object {
      "file": Object {
        "name": "file2.txt",
        "size": 9,
        "stream": [Function],
      },
      "progress": 0,
      "status": "validating",
    },
    Object {
      "file": Object {
        "name": "file3.txt",
        "size": 9,
        "stream": [Function],
      },
      "progress": 0,
      "status": "waiting",
    },
    Object {
      "file": Object {
        "name": "file4.txt",
        "size": 9,
        "stream": [Function],
      },
      "progress": 0,
      "status": "waiting",
    },
  ],
  "uploading": undefined,
  "validating": Object {
    "name": "file2.txt",
    "size": 9,
    "stream": [Function],
  },
}
`;

exports[`VALIDATION_FAILED 2`] = `
Object {
  "items": Array [
    Object {
      "err": "invalid encoding",
      "file": Object {
        "name": "file1.txt",
        "size": 9,
        "stream": [Function],
      },
      "progress": 0,
      "status": "validationFailed",
    },
    Object {
      "err": "invalid csv format",
      "file": Object {
        "name": "file2.txt",
        "size": 9,
        "stream": [Function],
      },
      "progress": 0,
      "status": "validationFailed",
    },
    Object {
      "file": Object {
        "name": "file3.txt",
        "size": 9,
        "stream": [Function],
      },
      "progress": 0,
      "status": "validating",
    },
    Object {
      "file": Object {
        "name": "file4.txt",
        "size": 9,
        "stream": [Function],
      },
      "progress": 0,
      "status": "waiting",
    },
  ],
  "uploading": undefined,
  "validating": Object {
    "name": "file3.txt",
    "size": 9,
    "stream": [Function],
  },
}
`;

exports[`VALIDATION_SUCCEEDED 1`] = `
Object {
  "items": Array [
    Object {
      "file": Object {
        "name": "file1.txt",
        "size": 9,
        "stream": [Function],
      },
      "progress": 0,
      "stats": Object {
        "columns": Array [
          "id",
          "name",
          "email",
        ],
        "hash": "hash",
        "rowCount": 1000,
      },
      "status": "uploading",
    },
    Object {
      "file": Object {
        "name": "file2.txt",
        "size": 9,
        "stream": [Function],
      },
      "progress": 0,
      "status": "validating",
    },
    Object {
      "file": Object {
        "name": "file3.txt",
        "size": 9,
        "stream": [Function],
      },
      "progress": 0,
      "status": "waiting",
    },
    Object {
      "file": Object {
        "name": "file4.txt",
        "size": 9,
        "stream": [Function],
      },
      "progress": 0,
      "status": "waiting",
    },
  ],
  "uploading": Array [
    Object {
      "file": Object {
        "name": "file1.txt",
        "size": 9,
        "stream": [Function],
      },
      "progress": 0,
      "stats": Object {
        "columns": Array [
          "id",
          "name",
          "email",
        ],
        "hash": "hash",
        "rowCount": 1000,
      },
      "status": "uploading",
    },
  ],
  "validating": Object {
    "name": "file2.txt",
    "size": 9,
    "stream": [Function],
  },
}
`;

exports[`VALIDATION_SUCCEEDED 2`] = `
Object {
  "items": Array [
    Object {
      "file": Object {
        "name": "file1.txt",
        "size": 9,
        "stream": [Function],
      },
      "progress": 0,
      "stats": Object {
        "columns": Array [
          "id",
          "name",
          "email",
        ],
        "hash": "hash",
        "rowCount": 1000,
      },
      "status": "uploading",
    },
    Object {
      "file": Object {
        "name": "file2.txt",
        "size": 9,
        "stream": [Function],
      },
      "progress": 0,
      "stats": Object {
        "columns": Array [
          "id",
          "name",
          "email",
        ],
        "hash": "hash",
        "rowCount": 1000,
      },
      "status": "uploading",
    },
    Object {
      "file": Object {
        "name": "file3.txt",
        "size": 9,
        "stream": [Function],
      },
      "progress": 0,
      "status": "validating",
    },
    Object {
      "file": Object {
        "name": "file4.txt",
        "size": 9,
        "stream": [Function],
      },
      "progress": 0,
      "status": "waiting",
    },
  ],
  "uploading": Array [
    Object {
      "file": Object {
        "name": "file1.txt",
        "size": 9,
        "stream": [Function],
      },
      "progress": 0,
      "stats": Object {
        "columns": Array [
          "id",
          "name",
          "email",
        ],
        "hash": "hash",
        "rowCount": 1000,
      },
      "status": "uploading",
    },
    Object {
      "file": Object {
        "name": "file2.txt",
        "size": 9,
        "stream": [Function],
      },
      "progress": 0,
      "stats": Object {
        "columns": Array [
          "id",
          "name",
          "email",
        ],
        "hash": "hash",
        "rowCount": 1000,
      },
      "status": "uploading",
    },
  ],
  "validating": Object {
    "name": "file3.txt",
    "size": 9,
    "stream": [Function],
  },
}
`;

exports[`VALIDATION_SUCCEEDED 3`] = `
Object {
  "items": Array [
    Object {
      "file": Object {
        "name": "file1.txt",
        "size": 9,
        "stream": [Function],
      },
      "progress": 0,
      "stats": Object {
        "columns": Array [
          "id",
          "name",
          "email",
        ],
        "hash": "hash",
        "rowCount": 1000,
      },
      "status": "uploading",
    },
    Object {
      "file": Object {
        "name": "file2.txt",
        "size": 9,
        "stream": [Function],
      },
      "progress": 0,
      "stats": Object {
        "columns": Array [
          "id",
          "name",
          "email",
        ],
        "hash": "hash",
        "rowCount": 1000,
      },
      "status": "uploading",
    },
    Object {
      "file": Object {
        "name": "file3.txt",
        "size": 9,
        "stream": [Function],
      },
      "progress": 0,
      "stats": Object {
        "columns": Array [
          "id",
          "name",
          "email",
        ],
        "hash": "hash",
        "rowCount": 1000,
      },
      "status": "uploading",
    },
    Object {
      "file": Object {
        "name": "file4.txt",
        "size": 9,
        "stream": [Function],
      },
      "progress": 0,
      "status": "validating",
    },
  ],
  "uploading": Array [
    Object {
      "file": Object {
        "name": "file1.txt",
        "size": 9,
        "stream": [Function],
      },
      "progress": 0,
      "stats": Object {
        "columns": Array [
          "id",
          "name",
          "email",
        ],
        "hash": "hash",
        "rowCount": 1000,
      },
      "status": "uploading",
    },
    Object {
      "file": Object {
        "name": "file2.txt",
        "size": 9,
        "stream": [Function],
      },
      "progress": 0,
      "stats": Object {
        "columns": Array [
          "id",
          "name",
          "email",
        ],
        "hash": "hash",
        "rowCount": 1000,
      },
      "status": "uploading",
    },
    Object {
      "file": Object {
        "name": "file3.txt",
        "size": 9,
        "stream": [Function],
      },
      "progress": 0,
      "stats": Object {
        "columns": Array [
          "id",
          "name",
          "email",
        ],
        "hash": "hash",
        "rowCount": 1000,
      },
      "status": "uploading",
    },
  ],
  "validating": Object {
    "name": "file4.txt",
    "size": 9,
    "stream": [Function],
  },
}
`;

exports[`VALIDATION_SUCCEEDED 4`] = `
Object {
  "items": Array [
    Object {
      "file": Object {
        "name": "file1.txt",
        "size": 9,
        "stream": [Function],
      },
      "progress": 0,
      "stats": Object {
        "columns": Array [
          "id",
          "name",
          "email",
        ],
        "hash": "hash",
        "rowCount": 1000,
      },
      "status": "uploading",
    },
    Object {
      "file": Object {
        "name": "file2.txt",
        "size": 9,
        "stream": [Function],
      },
      "progress": 0,
      "stats": Object {
        "columns": Array [
          "id",
          "name",
          "email",
        ],
        "hash": "hash",
        "rowCount": 1000,
      },
      "status": "uploading",
    },
    Object {
      "file": Object {
        "name": "file3.txt",
        "size": 9,
        "stream": [Function],
      },
      "progress": 0,
      "stats": Object {
        "columns": Array [
          "id",
          "name",
          "email",
        ],
        "hash": "hash",
        "rowCount": 1000,
      },
      "status": "uploading",
    },
    Object {
      "file": Object {
        "name": "file4.txt",
        "size": 9,
        "stream": [Function],
      },
      "progress": 0,
      "stats": Object {
        "columns": Array [
          "id",
          "name",
          "email",
        ],
        "hash": "hash",
        "rowCount": 1000,
      },
      "status": "validationSucceeded",
    },
  ],
  "uploading": Array [
    Object {
      "file": Object {
        "name": "file1.txt",
        "size": 9,
        "stream": [Function],
      },
      "progress": 0,
      "stats": Object {
        "columns": Array [
          "id",
          "name",
          "email",
        ],
        "hash": "hash",
        "rowCount": 1000,
      },
      "status": "uploading",
    },
    Object {
      "file": Object {
        "name": "file2.txt",
        "size": 9,
        "stream": [Function],
      },
      "progress": 0,
      "stats": Object {
        "columns": Array [
          "id",
          "name",
          "email",
        ],
        "hash": "hash",
        "rowCount": 1000,
      },
      "status": "uploading",
    },
    Object {
      "file": Object {
        "name": "file3.txt",
        "size": 9,
        "stream": [Function],
      },
      "progress": 0,
      "stats": Object {
        "columns": Array [
          "id",
          "name",
          "email",
        ],
        "hash": "hash",
        "rowCount": 1000,
      },
      "status": "uploading",
    },
  ],
}
`;

exports[`empty state 1`] = `
Object {
  "items": Array [],
}
`;
